odxtools/diaglayer.py

```diff
@@ -454,49 +455,115 @@ def get_communication_parameter(self, name: str) \
 
         return cps[0]
 
+    def _get_comparam_value(self,
+                            com_param : CommunicationParameterRef) \
+            -> Optional[str]:
+        """Retrieve the value of a simple communication parameter
+
+        This takes the default value of the comparam (if any) into
+        account.
+        """
+
+        assert isinstance(com_param.comparam, Comparam)
+
+        result: Any = None
+        if com_param.value:
+            result = com_param.value
+        else:
+            result = com_param.comparam.physical_default_value
+
+        assert isinstance(result, str)
+
+        return result
+
+    def _get_comparam_subvalue(self,
+                               com_param : CommunicationParameterRef,
+                               subparam_name: str) \
+            -> Optional[str]:
+        """Retrieve the value of a complex communication parameter's sub-parameter by name
+
+        This takes the default value of the comparam (if any) into
+        account.
+        """
+        comparam_spec = com_param.comparam
+        assert isinstance(comparam_spec, ComplexComparam)
+
+        value_list = com_param.value
+        if not isinstance(value_list, list):
+            warnings.warn(f"The values of complex communication parameter "
+                          f"'{com_param.short_name}' are not specified "
+                          f"correctly.", OdxWarning)
+            return None
+
+        name_list = [ cp.short_name for cp in comparam_spec.comparams ]
+        try:
+            idx = name_list.index(subparam_name)
+        except ValueError:
+            warnings.warn(f"Communication parameter '{com_param.short_name}' "
+                          f"does not specify a '{subparam_name}' sub-parameter.",
+                          OdxWarning)
+            return None
+
+        result = value_list[idx]
+        if not result and \
+           (default_values := comparam_spec.complex_physical_default_value):
+            result = default_values[idx]
+        assert isinstance(result, str)
+
+        return result
+
     def get_receive_id(self) -> Optional[int]:
         """CAN ID to which the ECU listens for diagnostic messages"""
         com_param = self.get_communication_parameter("CP_UniqueRespIdTable")
         if com_param is None:
             return None
-        value = com_param.value
-        if not isinstance(value, dict):
+
+        result = self._get_comparam_subvalue(com_param, "CP_CanPhysReqId")
```

Comment:
shouldn't `_get_comparam_subvalue` return an int if its DOP says?

Reply:
the `CommunicationParameterRef` class stores all values as strings ([[simple](https://github.com/mercedes-benz/odxtools/blob/main/odxtools/communicationparameter.py#L85)] [[complex](https://github.com/mercedes-benz/odxtools/blob/18c138ac29085f6dcd680b1a6b6d09e8a2629893/odxtools/comparam_subset.py#L31)])

This is an artifact from the time when the comparam specs where not yet internalized and it should probably be changed. (But not in this PR, because it has become a bit unwieldy already...)